src/app/archive/lib/processor.ml

```diff
@@ -3303,6 +3333,776 @@ module Block = struct
+            ; block_winner_id = Map.find_exn public_key_ids block.block_winner
+            ; last_vrf_output
+            ; snarked_ledger_hash_id =
+                Map.find_exn ledger_hash_ids block.snarked_ledger_hash
+            ; staking_epoch_data_id =
+                Map.find_exn epoch_ids
+                  (epoch_data_to_repr block.staking_epoch_data
+                     ~find_ledger_hash_id:(Map.find_exn ledger_hash_ids) )
+            ; next_epoch_data_id =
+                Map.find_exn epoch_ids
+                  (epoch_data_to_repr block.next_epoch_data
+                     ~find_ledger_hash_id:(Map.find_exn ledger_hash_ids) )
+            ; min_window_density =
+                block.min_window_density |> Mina_numbers.Length.to_uint32
+                |> Unsigned.UInt32.to_int64
+            ; sub_window_densities =
+                block.sub_window_densities
+                |> List.map ~f:(fun length ->
+                       Mina_numbers.Length.to_uint32 length
+                       |> Unsigned.UInt32.to_int64 )
+                |> Array.of_list
+            ; total_currency = Currency.Amount.to_string block.total_currency
+            ; ledger_hash = block.ledger_hash |> Ledger_hash.to_base58_check
+            ; height = block.height |> Unsigned.UInt32.to_int64
+            ; global_slot_since_hard_fork =
+                block.global_slot_since_hard_fork
+                |> Mina_numbers.Global_slot_since_hard_fork.to_uint32
+                |> Unsigned.UInt32.to_int64
+            ; global_slot_since_genesis =
+                block.global_slot_since_genesis
+                |> Mina_numbers.Global_slot_since_genesis.to_uint32
+                |> Unsigned.UInt32.to_int64
+            ; protocol_version_id =
+                Map.find_exn protocol_version_ids block.protocol_version
+            ; proposed_protocol_version_id =
+                Option.map block.proposed_protocol_version
+                  ~f:(Map.find_exn protocol_version_ids)
+            ; timestamp = Block_time.to_string_exn block.timestamp
+            ; chain_status = Chain_status.to_string block.chain_status
+            } )
+      in
+      let%map ids =
+        bulk_insert typ partial_missing_blocks ~table:table_name
+          ~fields:Fields.names
+      in
+      State_hash.Map.of_alist_exn
+      @@ List.zip_exn
+           (List.map missing_blocks ~f:(fun { state_hash; _ } -> state_hash))
+           ids
+    in
+    let block_ids =
+      let check_conflict ~key:_ conflict =
+        match conflict with
+        | `Both _ ->
+            failwith "duplicate data loaded during differential insertion"
+        | `Left x | `Right x ->
+            Some x
+      in
+      Map.merge existing_block_ids new_block_ids ~f:check_conflict
+      |> Map.merge external_block_ids ~f:check_conflict
+    in
+    let%bind () =
+      (* filter out direct descendants of the genesis block, as those maintain the NULL parent reference *)
```

Comment:
Should this be the last step? What if the `berkeley_migration` app crashed after this step is finished but the commands are not being added? Then the database would be in an inconsistent state where blocks are being added by the associated commands are missing.

Reply:
Yes, I explain this in the description of the PR. We delete any partially written blocks on bootup of the migration process to account for this issue.